runtime/globroots.c

```diff
@@ -198,70 +208,85 @@ CAMLexport void caml_remove_global_root(value *r)
   caml_delete_global_root(&caml_global_roots, r);
 }
 
+enum gc_root_class {
+  YOUNG,
+  OLD,
+  UNTRACKED
+};
+
+static enum gc_root_class classify_gc_root(value v)
+{
+  if(!Is_block(v)) return UNTRACKED;
+  if(Is_young(v)) return YOUNG;
+  if(Is_in_heap(v)) return OLD;
+  return UNTRACKED;
+}
+
 /* Register a global C root of the generational kind */
 
 CAMLexport void caml_register_generational_global_root(value *r)
 {
-  value v = *r;
   CAMLassert (((intnat) r & 3) == 0);  /* compact.c demands this (for now) */
-  if (Is_block(v)) {
-    if (Is_young(v))
+
+  switch(classify_gc_root(*r)) {
+    case YOUNG:
       caml_insert_global_root(&caml_global_roots_young, r);
-    else if (Is_in_heap(v))
+      break;
+    case OLD:
       caml_insert_global_root(&caml_global_roots_old, r);
+      break;
+    case UNTRACKED: break;
   }
 }
 
 /* Un-register a global C root of the generational kind */
 
 CAMLexport void caml_remove_generational_global_root(value *r)
 {
-  value v = *r;
-  if (Is_block(v)) {
-    if (Is_in_heap_or_young(v))
-      caml_delete_global_root(&caml_global_roots_young, r);
-    if (Is_in_heap(v))
+  switch(classify_gc_root(*r)) {
+    case OLD:
       caml_delete_global_root(&caml_global_roots_old, r);
+      /* Fallthrough: the root can be in the young list while actually
+         being in the major heap. */
+    case YOUNG:
+      caml_delete_global_root(&caml_global_roots_young, r);
+      break;
+    case UNTRACKED: break;
   }
 }
 
 /* Modify the value of a global C root of the generational kind */
 
 CAMLexport void caml_modify_generational_global_root(value *r, value newval)
 {
-  value oldval = *r;
-
-  /* It is OK to have a root in roots_young that suddenly points to
-     the old generation -- the next minor GC will take care of that.
-     What needs corrective action is a root in roots_old that suddenly
-     points to the young generation. */
-  if (Is_block(newval) && Is_young(newval) &&
-      Is_block(oldval) && Is_in_heap(oldval)) {
-    caml_delete_global_root(&caml_global_roots_old, r);
-    caml_insert_global_root(&caml_global_roots_young, r);
-  }
-  /* PR#4704 */
-  else if (!Is_block(oldval) && Is_block(newval)) {
-    /* The previous value in the root was unboxed but now it is boxed.
-       The root won't appear in any of the root lists thus far (by virtue
-       of the operation of [caml_register_generational_global_root]), so we
-       need to make sure it gets in, or else it will never be scanned. */
-    if (Is_young(newval))
-      caml_insert_global_root(&caml_global_roots_young, r);
-    else if (Is_in_heap(newval))
-      caml_insert_global_root(&caml_global_roots_old, r);
+  /* See PRs #4704, #607 and #8656 */
+  switch(classify_gc_root(newval)) {
+    case YOUNG:
+      switch(classify_gc_root(*r)) {
+        case OLD:
+          caml_delete_global_root(&caml_global_roots_old, r);
+          /* Fallthrough */
```

Comment:
Contrary to the one above, I'm not fond of this fallthrough here which is purely for code factorization, but does not make much sense to me. I would recommend to either duplicate the `caml_insert` call -- no problem with that, or have a structure of the form

```c
if (OLD) ...;
if !(YOUNG) ...;
```

which is more explicit about why things are shared. (But that requires declaring an intermediate variable, so it is different from your style in this function.)

Reply:
Actually I think just using `default:` instead of `case UNTRACKED:` would be clear enough; `default` means `non-YOUNG` here.

Reply:
New version pushed. I still prefer the `if` version. Your intiution for `default` is rather weird.

Reply:
Yes, the new version is nice! Thanks.